File: src/main/java/com/sothawo/springdataelastictest/presidents/PresidentHandler.java
package com.sothawo.springdataelastictest.presidents;

import com.sothawo.springdataelastictest.ResourceNotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.elasticsearch.core.SearchHit;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

import static org.springframework.web.reactive.function.server.ServerResponse.*;

@Component
class PresidentHandler {
    private static final Logger LOGGER = LoggerFactory.getLogger(PresidentHandler.class);

    private final PresidentService service;
    PresidentHandler(PresidentService service){
        this.service = service;
    }

    public Mono<ServerResponse> load(ServerRequest request) {
        return ok().body(service.load(), President.class);
    }

    public Mono<ServerResponse> byId(ServerRequest request) {
        var id = request.pathVariable("id");
        var presidentMono = service.byId(id).switchIfEmpty(Mono.error(new ResourceNotFoundException()));
        return ok().body(presidentMono, President.class);
    }

    public Mono<ServerResponse> searchByTerm(ServerRequest request) {
        var year = Integer.parseInt(request.pathVariable("year"));
        return ok().body(service.searchByTerm(year), SearchHit.class);
    }

    public Mono<ServerResponse> searchByName(ServerRequest request) {
        var name = request.pathVariable("name");
        Boolean requestCache = null;
        var queryParam = request.queryParam("requestCache");
        if (queryParam.isPresent()) {
            requestCache = Boolean.parseBoolean(queryParam.get());
        }

        var searchHits = service.searchByName(name, requestCache);

        return ok().body(searchHits, SearchHit.class);
    }
}

File: src/main/kotlin/com/sothawo/springdataelastictest/presidents/PresidentHandler.kt
package com.sothawo.springdataelastictest.presidents

import org.springframework.data.elasticsearch.core.SearchHit
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono
import kotlin.let

@Component
class PresidentHandler(
    private val service: PresidentService
) {
    suspend fun load(request: ServerRequest): ServerResponse =
        ServerResponse.ok().bodyAndAwait(service.load())

    suspend fun byId(request: ServerRequest): ServerResponse {
        val id = request.pathVariable("id")
        return ServerResponse.ok().bodyValueAndAwait(service.byId(id))
    }

    suspend fun searchByTerm(request: ServerRequest): ServerResponse {
        val year = request.pathVariable("year").toInt()
        return ServerResponse.ok().json().bodyAndAwait(service.searchByTerm(year))
    }

    suspend fun searchByName(request: ServerRequest): ServerResponse {
        val name = request.pathVariable("name")
        val requestCache = request.queryParamOrNull("requestCache")?.let { it -> it.toBoolean() }

        return ServerResponse.ok().bodyAndAwait(service.searchByName(name, requestCache))
    }
}
